### ISSUE TITLE
Partial inference specifically when declaring variables

### ISSUE THREAD

**Issue by schicks** [microsoft/TypeScript#46685]
# Suggestion
I frequently end up with the following pattern in my code.
```typescript
type Item = {
    some: 'internal'
    structure: boolean
}

const untypedMapOfItems = {
  idA: {
    some: 'internal',
    structure: true
  },
  idB: {
      some: 'internal',
      structure: false
  }
}

const mapOfItems: {[Key in keyof typeof untypedMapOfItems]: Item} = untypedMapOfItems
```

The goal of this pattern is to avoid having to repeat the ids in some type, so that they will just be carried through to the type of `mapOfItems`, while still validating that the values of `untypedMapOfItems` have to be `Item`. This is very useful to me in that it maintains type safety while avoiding boilerplate types, where we need to write a series of ids twice. However, it feels pretty boilerplatey itself, and I would like a shorthand to carry through some _part_ of the type of an assignment while still maintaining some invariants. Something like;
```typescript
type Item = {
    some: 'internal'
    structure: boolean
}

const mapOfItems: {[Key in (infer K)]: Item} = {
  idA: {
    some: 'internal',
    structure: true
  },
  idB: {
      some: 'internal',
      structure: false
  }
}
```

This would mirror the syntax from conditional types, and effectively be a shorthand for

```typescript
type Item = {
    some: 'internal'
    structure: boolean
}

const untypedMapOfItems = {
  idA: {
    some: 'internal',
    structure: true
  },
  idB: {
      some: 'internal',
      structure: false
  }
}

const mapOfItems: typeof untypedMapOfItems extends {[Key in infer K]: Item} ? {[Key in K]: Item} : never = untypedMapOfItems
```

Both my current go to solution and the conditional type analogy example currently produce relatively hard to understand type errors or produce type errors in strange places. Hopefully this new syntax could produce type errors more similar to the following code, but keep the tighter type.
```typescript
type Item = {
    some: 'internal'
    structure: boolean
}

const untypedMapOfItems: {[Key in string]: Item} = {
  idA: {
    some: 'internal',
    structure: true
  },
  idB: {
      some: 'internal',
      structure: false
  }
}
```

This is a [sandbox](https://www.typescriptlang.org/play?#code/C4TwDgpgBAksEFsoF4oG8CwAoKuoGcB7BCALigHIBLAO3gCcaBDAGwuzwOHoFcBjYD3pkoAI0KEWEJjWwBfbNj6Ea+YFB51wEACYBZJmADyAMziJ8KdByhUdAQXJo5AGht2AQk5uciJctR0EIysFG44nFy8AkIiJqz4EDYKWCnYAPTpUHxCwnRQYEzADDQF9IQ6-BCWwkxEzKJSUKCQUMHl9LalTARULCwgBSxMfElYyqrqwgAq2vqGpuYI+E4A2gDSEIO0UADWW4QmzdqHGlqQ88Zm8MsAuuRLclaaLboGV0v4iliZUDKshAA5oNgIRsiodFRgFQVKxjpBLGBypVRpZREwdPDoO1CJ0dj18H0BkMRmMJmooAgFtcLORXqcXnN3osbpYIAAPeA0HSWNAbLZdLomYJQdb3WA3J4AfnQ-O2pTFD0lUHINAgADcRahGRdmTTlt8MllgAALKiI5FVSw9QESTGvNr0DqCm12kmjFxiHjqYJ1arNE3QfCQPhUExUPhYqCnU3QfYgL7jFQUtT0WiA2a66mfNabeVRdPix5WTAROyOdCudw6LzWCK+YgiQIlULhSJRfiCYTkeIsRLJeRAA) exploring the various examples that are currently valid typescript.
## 🔍 Search Terms

inference, partial inference, infer syntax, conditional types, boilerplate reduction

## ✅ Viability Checklist
My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by andrewbranch** [comment]
This looks like a different solution to the same problem that https://github.com/microsoft/TypeScript/issues/7481 tries to solve.

**Comment by andrewbranch** [comment]
Also reminds me of https://github.com/apple/swift-evolution/blob/main/proposals/0315-placeholder-types.md

**Comment by schicks** [comment]
Placeholder types do seem like a pretty close comparison. I do think #7481 would _largely_ address this for me, though I've never really felt the need for this pattern other than at assignment, so I have a slight preference to the syntax I suggest above.

**Comment by jcalz** [comment]
Is this significantly different from #38349 (which was closed as a duplicate of #7481 also)?
